# 电商页面A/B测试  
## 背景  
   某电商公司非常注重自己的**落地页**设计，希望通过改进设计来提高**转化率**。以往该公司全年转化率平均在13%左右，现在希望设计的新页面能够带来更高的转化率，希望新页面的转化率能有2%的提升，达到15%。在正式推出新页面之前，该公司希望通过AB测试在小范围的用户中进行测试，以确保新页面的效果能够达到预期目标。  
## ABtest  
通常包括以下流程：  
### 1. 实验改动点确定  
在ABtest开始前，需要和业务部门确定准确的实验改动点，即实验变量。本项目中实验变量为落地页的不同设计风格。  

### 2. 确定实验目标和衡量指标  
本实验的目标为通过AB测试确定新的页面设计是否能够提高2%的转化率，衡量指标为转化率。  

### 3.提出假设  
   假设是对于某个特定变化我们所期望的结果，也是后续实验的基础，我们需要在后续实验中通过数据验证这个假设是否成立。如果验证成立，我们可以将这个变化推广到全部用户。如果验证不成立，则需要继续优化这个假设或者放弃这个修改方案，以寻找更好的变化。  
  在这个实验中，我们希望新页面可以提升2%的转化率，原则上我们应选择右侧单尾检验，因为我们的假设是新页面的转化率要大于旧页面的转化率。但是，在本案例中，我们并不能确定新页面的性能一定比当前的页面更好，转化率更高只是我们的希望。我们要先验证它们存在差异，所以，这里选择双尾检验。  
  如果实验的目的是验证新设计的落地页转化率是否明显高于旧设计，就可以选择单尾检验。但如果不确定两种设计是否存在差异，则应选择双尾检验。  
			
  通常，将我们期望达到的结果放到备择假设上（拒绝原假设，接受备择假设更有说服力），在本实验中，提出假设如下：  
				H0：P0=P1  
				H1：P0≠P1  
  其中，P0表示旧版页面设计的转化率，P1表示新版页面设计的转化率  

### 4.实验分组（流量分割）  
  这一步目的是确保实验组和对照组的样本分布均匀且具有随机性，从而最大程度地减少其他因素对实验结果的干扰。具体来说，合理的样本分割应遵循随机化原则，使实验组和对照组在性别、年龄、地区、设备类型、活跃程度等关键特征上尽可能一致，以避免潜在的系统性偏差。通过这样的方法，可以确保在实验期间，只有实验改动点这一因素对核心指标产生影响，而其他无关变量被有效控制。  
	此外，合理的随机分组有助于避免辛普森悖论的发生。辛普森悖论指的是，当数据被分组分析时，总体趋势可能会被分组内的特定趋势掩盖或逆转，从而得出错误的结论。这种现象往往是由于样本分布不均匀导致的。  
		
  本次实验中，分组 如下：  
  * 对照组：用户看到旧版的落地页。  
  *   实验组：用户看到新版的落地页。  
  
  为了便于计算转化率，对用户购买行为进行记录：  
	         	0：代表用户在测试期间没有购买产品  
				    1：代表用户在测试期间购买了产品  
  后续通过求平均值，得到每个组的转化率。  
				
### 5. 计算实验最小样本量和实验周期  
* **最小样本量**  
AB测试是用小样本来推断总体，总体的分布有多种可能 ，可能是正态分布，也可能是偏态分布。  
根据中心极限定理：在样本量足够的的情况下（>30）,样本均值的分布，趋近于正态分布，正态分布曲线由均值和方差决定，该分布均值的均值E( ̅x)趋近于总体均值u。  
可以通过抽样，借助正态分布来估计置信区间，实现参数检验（t检验等）。  

样本量越大，估计就越准确，但成本也就越高。如何选定合适的最小的样本量？  
最小样本量计算公式如下：  
													![Image Name](https://cdn.kesci.com/upload/sy1gvj7hu6.png?imageView2/0/w/960/h/960)  
						
n为对照组和实验组各组所需要的样本量，总样本即为2n  
σ为标准差  
Δ为效应大小，指预期的两组差异幅度，在本项目中，Δ等于实验组的转化率减去对照组的转化率  
α为犯第一类错误的概率，即弃真错误的概率。  
𝛼=𝑃(拒绝 𝐻0∣𝐻0 为真)  
β为犯第二类错误的概率，即纳伪错误的概率。  
β=P(接受 H0∣H0 为假)  
一般情况下我们设置：  
显著性水平：α=0.05（或0.01）  
统计功效（1−β）：β=0.2，即指检测到真实差异的概率，通常设为 0.8。在此案例中就是，如果新版落地页真的比旧版转换率要高，该测试有80%的概率能检测出这个状况。  
当衡量指标为比率类指标时，标准差计算公式为：  
											![Image Name](https://cdn.kesci.com/upload/sy1huhxk4p.png?imageView2/0/w/960/h/960)  
其中，PA和PB分别是对照组和实验组的观测值。在此案例中PA=13%，PB=15%  
代入数据可求得，本实验 的最小 样本量为 N = 4716。  



In [140]:
(0.13*(1-0.13)+0.15*(1-0.15))/(0.15-0.13)**2*(1.96+0.84)**2

4715.760000000005


**借助在线工具求最小样本量：[https://www.evanmiller.org/ab-testing/sample-size.html](https://www.evanmiller.org/ab-testing/sample-size.html)**  
																							![Image Name](https://cdn.kesci.com/upload/sy1i85aeul.png?imageView2/0/w/960/h/960)  
		



**使用python完成最小样本量计算：**

In [141]:
import numpy as np 
import pandas as pd 
import scipy.stats as stats 
import statsmodels.stats.api as sms 
import matplotlib as mpl 
import matplotlib.pyplot as plt 
import seaborn as sns


In [142]:
#计算效应大小，该函数用于比较两个比例之间的标准化差异
effect_size = sms.proportion_effectsize(0.13, 0.15)

In [143]:
effect_size

-0.0576728617308947

In [144]:
#基于效应大小、功效、显著性水平等参数，计算每组所需的样本量
required_n = sms.NormalIndPower().solve_power( 
    effect_size, 
    power=0.8, #统计功效
    alpha=0.05, #显著水平
    ratio=1 ,
    )

In [145]:
required_n

4719.4740575998185

In [146]:
np.ceil(required_n) #向上取整

4720.0

底层实际是针对比例差异的样本计算公式：  
											![Image Name](https://cdn.kesci.com/upload/sy1j7v3dy9.png?imageView2/0/w/960/h/960)  


* **实验周期**  

结合样本量和日均活跃用户数（DAU）计算实验周期。  
根据上面最小样本量的计算，我们知道此次AB测试至少需要9440个用户参与测试，假如该落地页以往每天的平均浏览人数为1000，则实验周期至少需要的天数为：  
																																											9440/1000=9.4,约10天  
### 6. 进行灰度测试  
在正式开启ABtest之前，为了确保实验的改动不会对系统或用户体验造成巨大的异常影响，通常需要先选取一小部分用户样本进行灰度测试，旨在通过对少量用户的观察，评估实验改动点的风险和稳定性。  
通过灰度测试，可以验证实验的功能是否正常，用户体验是否符合预期，以及核心指标是否出现重大波动。如果灰度测试结果良好且风险可控，便可逐步扩大分流比例，最终进入正式的ABtest阶段；否则，需对实验方案或改动点进行调整，避免全面上线后带来不可控的负面影响。  
### 7. AB测试上线和数据收集  
根据确定好的实验方案，正式进行AB测试，同时通过特定的方法对核心指标进行监控，记录实验数据，便于后续分析。  


### 8. 数据分析及假设检验  
假设我们已经收集好了数据（这里使用Kaggle上的电商页面AB测试数据集）  
#### 导入数据集

In [147]:
data = pd.read_csv("/home/mw/input/yijunyinzi2648/ab_data.csv")
data.head()

user_id                   timestamp      group landing_page  converted
0   851104  2017-01-21 22:11:48.556739    control     old_page          0
1   804228  2017-01-12 08:01:45.159739    control     old_page          0
2   661590  2017-01-11 16:55:06.154213  treatment     new_page          0
3   853541  2017-01-08 18:28:03.143765  treatment     new_page          0
4   864975  2017-01-21 01:52:26.210827    control     old_page          1

该数据集共有四个字段:  
user_id:用户的id  
timestamp:用户访问页面的时间  
group:用户属于 哪个组（control为对照组，treatment为实验组）  
landing_page:用户访问的是旧版落地页 还是新版落地页  
converted:是否成功转化（0代表未转化，1代表转化）

In [148]:
#查看数据的整体信息
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294478 entries, 0 to 294477
Data columns (total 5 columns):
user_id         294478 non-null int64
timestamp       294478 non-null object
group           294478 non-null object
landing_page    294478 non-null object
converted       294478 non-null int64
dtypes: int64(2), object(3)
memory usage: 11.2+ MB


一共有29万多条数据，目前看来没有缺失值，注意这里timestamp不是日期，而是文本类型，需要后续处理。

#### 数据清洗  
* **缺失值处理**

In [149]:
data.isnull().sum()   # 检查每一列是否有缺失值

user_id         0
timestamp       0
group           0
landing_page    0
converted       0
dtype: int64

* **重复值处理**

In [150]:
# 总体数据是否有重复
data.duplicated().sum()

0

In [151]:
# 用户id是否重复（这里我们是针对不同的用户进行研究，每个用户记录应该唯一 ）
data["user_id"].duplicated().sum()

3894

这里查找到3894条重复的用户记录，接下来进行处理：

In [152]:
# 查看重复的记录
data[data["user_id"].duplicated()]["user_id"]

2656      698120
2893      773192
7500      899953
8036      790934
10218     633793
11215     839784
11428     926754
12858     793494
13262     892450
14487     883654
15112     781656
15358     790456
15527     886965
16677     870096
18497     745943
18679     700810
20019     664347
20250     779408
20798     807355
21116     667654
21167     766921
22367     635809
23996     848099
25451     634971
25942     788011
26953     646066
27162     920614
27174     666555
27260     894409
27389     664025
           ...  
293443    738761
293449    715367
293480    638376
293530    934040
293544    817753
293548    797335
293568    704024
293604    648354
293628    700036
293657    914482
293659    691336
293662    927109
293706    795519
293712    811222
293773    688144
293817    876037
293888    865405
293894    741581
293917    738357
293980    916033
293996    942612
294014    813406
294200    928506
294252    892498
294253    886135
294308    905197
294309    787083
294328    6415

In [153]:
# 查看某一位重复的用户
data[data["user_id"]==698120]

user_id                   timestamp    group landing_page  converted
988    698120  2017-01-22 07:09:37.540970  control     new_page          0
2656   698120  2017-01-15 17:13:42.602796  control     old_page          0

发现这些重复的记录出现了某个用户既进入新版又进入旧版页面、用户的重复记录对应的是不同分组、分组与看到的页面不对应等情况，下一步应该删除这些重复记录 。

In [154]:
repeat_id = data[data["user_id"].duplicated()]["user_id"]
# repeat_id
data_new = data[~data["user_id"].isin(repeat_id)]

In [155]:
data_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 286690 entries, 0 to 294477
Data columns (total 5 columns):
user_id         286690 non-null int64
timestamp       286690 non-null object
group           286690 non-null object
landing_page    286690 non-null object
converted       286690 non-null int64
dtypes: int64(2), object(3)
memory usage: 13.1+ MB


In [156]:
data_new["user_id"].duplicated().sum()

0

* **查找分组和落地页不匹配的数据**

In [157]:
# group和landing_page交叉统计查看是否一一匹配
pd.crosstab(data_new["group"],data_new["landing_page"])

landing_page  new_page  old_page
group                           
control              0    143293
treatment       143397         0

发现没有出现像重复的user_id记录中的分组和看到的页面不匹配的情况

* **处理时间，查看统计天数**

In [158]:
date_days=pd.to_datetime(data["timestamp"]).dt.strftime("%Y-%m-%d")
#也可以用字符串的split方法实现
# data_days = [i[0] for i in data["timestamp"].str.split(" ")]
# date_days
len(date_days.unique())

23

一共有23天的数据，满足之前计算的实验周期要求

####  抽样  
因为这里用的数据集很大，所以根据最小样本量（4720）抽样，这里每组抽取5000条样本。  
实际情况下，是不需要抽样的，因为是根据实现计算好的样本量来部署AB测试，收集到的数据基本都是要用上的。

In [159]:
required_n = 5000
control_sample = data_new[data_new["group"]=='control'].sample(n=required_n)
treatment_sample = data_new[data_new["group"]=='treatment'].sample(n=required_n)

In [160]:
test_data = pd.concat([control_sample,treatment_sample],ignore_index=True)
test_data

user_id                   timestamp      group landing_page  converted
0      804751  2017-01-22 02:07:31.385618    control     old_page          0
1      702067  2017-01-05 07:02:47.008599    control     old_page          1
2      633628  2017-01-17 15:07:19.883592    control     old_page          0
3      819938  2017-01-16 20:32:51.122348    control     old_page          0
4      828565  2017-01-03 23:01:10.756589    control     old_page          0
5      677951  2017-01-12 12:06:15.001918    control     old_page          0
6      741497  2017-01-12 08:10:08.745717    control     old_page          0
7      660856  2017-01-10 19:26:19.561787    control     old_page          0
8      835574  2017-01-07 23:27:05.772293    control     old_page          0
9      752814  2017-01-06 00:01:38.699571    control     old_page          1
10     693758  2017-01-07 19:40:48.572538    control     old_page          0
11     815586  2017-01-14 19:21:27.280126    control     old_page          0
12     842270  2017-01-18 20:43:03.183784    control     old_page          0
13     869435  2017-01-06 01:02:29.219224    control     old_page          0
14     795632  2017-01-05 10:21:47.321331    control     old_page          0
15     662549  2017-01-04 09:23:17.807982    control     old_page          0
16     705088  2017-01-10 05:33:33.994972    control     old_page          0
17     750118  2017-01-22 00:23:19.848671    control     old_page          1
18     700373  2017-01-15 02:34:08.529149    control     old_page          0
19     876654  2017-01-15 19:01:29.705254    control     old_page          0
20     664029  2017-01-04 21:19:56.145766    control     old_page          0
21     874763  2017-01-15 22:27:42.313324    control     old_page          1
22     631220  2017-01-17 22:29:52.657991    control     old_page          0
23     711755  2017-01-18 08:41:20.118542    control     old_page          0
24     791066  2017-01-23 19:44:00.899977    control     old_page          0
25     656279  2017-01-21 04:59:47.254059    control     old_page          0
26     674521  2017-01-12 12:46:18.001569    control     old_page          0
27     803768  2017-01-15 16:24:26.531680    control     old_page          0
28     828569  2017-01-04 21:28:03.653700    control     old_page          0
29     784210  2017-01-13 16:49:00.296603    control     old_page          0
...       ...                         ...        ...          ...        ...
9970   750937  2017-01-24 01:28:10.361378  treatment     new_page          0
9971   842077  2017-01-16 05:26:15.816412  treatment     new_page          0
9972   700413  2017-01-15 12:52:56.826290  treatment     new_page          1
9973   748711  2017-01-12 14:03:41.526167  treatment     new_page          0
9974   873485  2017-01-10 23:22:36.741225  treatment     new_page          0
9975   659882  2017-01-20 03:01:49.841370  treatment     new_page          0
9976   721500  2017-01-07 23:17:19.470758  treatment     new_page          0
9977   831330  2017-01-23 03:01:35.648231  treatment     new_page          1
9978   735825  2017-01-04 12:01:46.927030  treatment     new_page          0
9979   640566  2017-01-10 10:05:41.833815  treatment     new_page          0
9980   660572  2017-01-09 22:38:52.531248  treatment     new_page          0
9981   656408  2017-01-08 10:54:50.885001  treatment     new_page          0
9982   681701  2017-01-04 13:16:37.916640  treatment     new_page          0
9983   845749  2017-01-14 10:08:35.344614  treatment     new_page          0
9984   692120  2017-01-09 07:20:13.260422  treatment     new_page          0
9985   799845  2017-01-20 14:29:12.335283  treatment     new_page          0
9986   729226  2017-01-20 08:46:18.490833  treatment     new_page          0
9987   794130  2017-01-03 09:47:26.059025  treatment     new_page          0
9988   909428  2017-01-23 20:35:09.607875  treatment     new_page          0
9989   776963  2017-01-04 13:24:41.673863  treatment     new_page          0
99

In [161]:
pd.crosstab(test_data["group"],test_data["landing_page"])

landing_page  new_page  old_page
group                           
control              0      5000
treatment         5000         0

####  可视化  
对两组实验结果的转化率进行可视化

In [162]:
conversion_rates = test_data.groupby("group")["converted"].agg([np.mean,np.std])
conversion_rates

mean       std
group                      
control    0.1200  0.324994
treatment  0.1262  0.332108

In [163]:
conversion_rates.columns = ["conversion_rate","std_deviation"]
conversion_rates

conversion_rate  std_deviation
group                                    
control             0.1200       0.324994
treatment           0.1262       0.332108

In [164]:
# plt.rcParams['font.family'] = 'SimHei' # 设置中文字体为黑体
plt.rcParams['axes.unicode_minus'] = False  # 解决-的编码问题

plt.figure(figsize=(8,6),dpi=60)  # 创建图表并设置大小和清晰度

sns.barplot(x=test_data['group'], y=test_data['converted'], ci=False)   # 绘制柱状图

plt.ylim(0, 0.17) 
plt.title( 'Group Conversion Rate', fontsize=20) 
plt.xlabel('group', fontsize=15) 
plt.ylabel('conversion_rate', fontsize=15);

<Figure size 480x360 with 1 Axes>

可见，新旧两版落地页的转化率非常接近，相对旧版，新版的转化率高了0.8%  
那么我们可以说新版落地页的转化率更好吗？这样的差异在统计学上是显著的吗？还只是偶然的结果？  
接下来需要进行假设检验来进行客观的分析。

####  假设检验  
这里是对新旧版页面的转化率是否相等进行检验，属于对比例进行检验。样本量足够大，总体方差未知，所以采用Z检验的方法，构造Z统计量：  
Z=（p1-p2）/sqrt[p1×(1-p1)/n1+p2×(1-p2)/n2]

利用Python中的statsmodels.stats.proportion模块可以来计算P值和置信区间：

In [165]:
from statsmodels.stats.proportion import proportions_ztest,proportion_confint
# proportions_ztest需要传入每一组的数量和成功转化的数量
# 取出converted值
control_results = test_data[test_data["group"] == 'control']["converted"]
treatment_results = test_data[test_data["group"] == 'treatment']["converted"]

In [166]:
n_con = control_results.count()
n_treat = treatment_results.count()
nobs = [n_con,n_treat]
successes = [control_results.sum(),treatment_results.sum()]

In [167]:
nobs

[5000, 5000]

In [168]:
successes

[600, 631]

In [169]:
# 计算Z值和P值
z_stat,p_val = proportions_ztest(successes,nobs)

# 计算临界值
critical_value = stats.norm.ppf(1-0.025)

# 计算对照组和实验组95%置信区间
(lower_con,lower_treat),(upper_con,upper_treat) = proportion_confint(successes,nobs,alpha=0.05)

In [171]:
# 打印结果
print(f"p_value : {p_val:.3f}")
print(f"z_statistic : {z_stat:.2f}")
print(f"临界值为 : {critical_value:.2f}")
print("落入拒绝域拒绝原假设,两版本页面存在差异" if abs(z_stat) > critical_value else "接受原假设，两版本页面不存在差异") 
print(f"ci 95% for control group : [{lower_con:.3f},{upper_con:.3f}]")
print(f"ci 95% for control group : [{lower_treat:.3f},{upper_treat:.3f}]")

p_value : 0.345
z_statistic : -0.94
临界值为 : 1.96
接受原假设，两版本页面不存在差异
ci 95% for control group : [0.111,0.129]
ci 95% for control group : [0.117,0.135]


## 分析结果与建议  


从上面的输出结果可以看出，p值远高于显著水平0.05，并且z统计量的值也是在接受域，也就是说新旧两版页面的转化率并没有显著的不同。  

此外，从置信区间来看，相比于旧版页面，新版页面转化率的置信区间  
* 有一定的增加，但是差异非常微小  
* 上限接近13%，明显低于15%  

综上，新版页面的真实转化率更可能接近基线值13%，与旧版页面并无明显差异，说明当前的修改并不有效 ，下一步应该继续对页面设计进行其他方面的优化，也可以通过用户调研的方式探索更能满足用户需求的页面设计。